src/controller.rs

```diff
@@ -0,0 +1,239 @@
+            .set_ownership(self.client_uid, self.client_gid)
+            .set_job_dir(job_dir)
+            .spawn()?;
+
+        let job_id: Uuid = job.id().to_owned();
+        self.jobs.insert(job_id, job);
+
+        Ok(job_id)
+    }
+
+    pub async fn status(&self, job_id: Uuid) -> Result<runner::JobStatus, Error> {
+        let job = self.jobs.get(&job_id).ok_or(Error::JobNotFound(job_id))?;
+        Ok(job.status())
+    }
+
+    pub async fn stop(&mut self, job_id: Uuid) -> runner::Ack {
+        if let Some(job) = self.jobs.get(&job_id) {
+            job.cancel();
+        }
+
+        runner::Ack {}
+    }
+
+    pub async fn output(&self, job_id: Uuid) -> Result<Receiver<Result<LogMessage, Error>>, Error> {
+        let job = self.jobs.get(&job_id).ok_or(Error::JobNotFound(job_id))?;
+        let job_dir = job.job_dir();
+
+        let (tx, rx) = mpsc::channel(20);
+
+        if job.is_complete() {
+            Self::read_file(Fd::Out, job_dir, tx.clone()).await;
+            Self::read_file(Fd::Err, job_dir, tx).await;
+
+            Ok(rx)
+        } else {
+            let completion = job.subscribe();
+            Self::watch_file(Fd::Out, job_dir, completion.clone(), tx.clone()).await;
+            Self::watch_file(Fd::Err, job_dir, completion, tx).await;
+
+            Ok(rx)
+        }
+    }
+
+    async fn watch_file(
+        fd: Fd,
+        job_dir: &Path,
+        mut completion: watch::Receiver<bool>,
+        tx: mpsc::Sender<Result<LogMessage, Error>>,
+    ) {
+        let file_path = match fd {
+            Fd::Out => job_dir.join("out"),
+            Fd::Err => job_dir.join("err"),
+        };
+
+        tokio::spawn(async move {
+            let fd: i32 = fd.into();
+            let mut buf = [0u8; 512];
+            let mut file = File::open(file_path).await?;
+            let mut done = false;
+
+            loop {
+                tokio::select! {
+                    _ = completion.changed() => {
+                        // exit on next EOF;
+                        done = true;
+                    }
+                   outcome = file.read(&mut buf) => {
+                       match outcome {
+                           Ok(bytes_read) if bytes_read > 0 => {
+                               let msg = LogMessage { fd, output: buf[..bytes_read].to_vec() };
+                               if let Err(err) = tx.send(Ok(msg)).await {
+                                   error!("Failed to send read logs({})", err);
+                                   break;
+                               }
+                           }
+                           Ok(bytes_read) if bytes_read == 0 && done => break,
+                           Err(err) => {
+                               error!("Error reading from log file: {}", err);
+                               break;
+                           },
+                           _ => (),
+                       }
+                   }
+
+                }
```

Comment:
Spinning on read until bytes appear should be avoided.

Worth noting that it's acceptable and encouraged to just handle stdio entirely in-memory.  Buffering to a file is allowable but not required.